### PR TITLE
[DEV-3204] Render Speed Improvements

### PR DESCRIPTION
**Pull request by 00salmon**
## Description of the change

> EmbedWidget has been upgraded to allow the ability to load the widget into a hidden element (container) and be opened/loaded on command (like the popup widget)

> Popup widget updated to allow a custom trigger selector for opening the widget, and an `sq:refresh` event is also sent

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation or Development tools (readme, specs, tests, code formatting)

## Links

 - Jira issue number: https://saasquatch.atlassian.net/browse/DEV-3204
 - Process.st launch checklist: https://app.process.st/runs/DEV-3204-SquatchJS-Render-Speed-Improvements-puMMWmgVCQf_8-alCE9ErA

### Development

- [x] [Prettier](https://www.npmjs.com/package/prettier) was run
- [x] The behaviour changes in the pull request are covered by specs
- [ ] All tests related to the changed code pass in development

### Paperwork

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has a Jira number
- [x] This pull request has a Process.st launch checklist
- [ ] "Ready for review" label attached to the PR and reviewers pinged on Slack

### Code review 

- [ ] Changes have been reviewed by at least one other engineer
- [ ] Security impacts of this change have been considered

### CODE REVIEW

File: src/widgets/Widget.ts [hunk 1]
```diff
@@ -91,6 +91,7 @@ export default abstract class Widget {
       };
     }
 
+    console.log("SAM LOOK", this, this.analyticsApi);
```
Comment (00salmon) [inline src/widgets/Widget.ts:94]:
```suggestion
```

File: src/widgets/Widgets.ts [hunk 2]
```diff
@@ -239,6 +237,7 @@ export default class Widgets {
     let displayCTA = false;
     const opts = response.jsOptions || "";
 
+    console.log("params config??", { context });
```
Comment (00salmon) [inline src/widgets/Widgets.ts:240]:
```suggestion
```

File: demo/toolbar.tsx [hunk 3]
```diff
@@ -341,79 +352,174 @@ async function getCustomWidget(engagementMedium) {
...
+  });
 }
 
 function MockedWidgets(props) {
   const { versions } = props;
-
   const [engagementMedium, setEngagementMedium] = useState("EMBED");
+  const [usePreload, setUsePreload] = useState(false);
+  const [showWidget, setShowWidget] = useState(false);
+  // const [element, setElement] = useState(false);
```
Comment (00salmon) [inline demo/toolbar.tsx:365]:
```suggestion
```

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@saasquatch/squatch-js",
-  "version": "2.3.1",
+  "version": "2.3.2-23",
```
Comment (00salmon) [inline package.json:3]:
```suggestion
  "version": "2.3.2-23",
```
TODO: figure out what version we are releasing

File: src/widgets/EmbedWidget.ts [hunk 5]
```diff
@@ -14,22 +13,47 @@ const _log = debug("squatch-js:EMBEDwidget");
...
+      this.targetElement.frame = this.frame;
+      this.targetElement._loadEvent = this._loadEvent;
+      this.targetElement.analyticsApi = this.analyticsApi;
+    }
     const element =
       document.querySelector(selector) ||
       document.querySelector(".squatchembed");
 
-    if (!element)
+    if (element === undefined)
```
Comment (00salmon) [inline src/widgets/EmbedWidget.ts:34]:
```suggestion
    if (!element)
```

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@saasquatch/squatch-js",
-  "version": "2.3.1",
+  "version": "2.3.2-26",
```
Comment (DerekSiemens) [inline package.json:3]: This probably actually should be `2.4.0`, this is not really a patch

File: src/widgets/EmbedWidget.ts [hunk 7]
```diff
@@ -73,11 +85,39 @@ export default class EmbedWidget extends Widget {
...
 
+  // Un-hide if element is available and refresh data
+  open() {
+    if (!this.frame || !this.context.container)
+      return _log("no target element to open");
+    this.element.style.visibility = "unset";
+    this.element.style.height = "auto";
+    this.element.style["overflow-y"] = "auto";
+
+    this.frame?.contentDocument?.dispatchEvent(new CustomEvent("sq:refresh"));
```
Comment (DerekSiemens) [inline src/widgets/EmbedWidget.ts:104]: This should also be added to the popup widget

File: src/widgets/EmbedWidget.ts [hunk 8]
```diff
@@ -14,22 +13,35 @@ const _log = debug("squatch-js:EMBEDwidget");
...
 export default class EmbedWidget extends Widget {
-  element: Element;
+  element: HTMLElement;
 
   constructor(params: Params, selector = "#squatchembed") {
     super(params);
 
     const element =
+      params.context.container ||
       document.querySelector(selector) ||
```
[the remark this answers is not in the document]
Comment (DerekSiemens) [reply src/widgets/EmbedWidget.ts:23]: We should probably add supporting it into this PR

File: src/widgets/EmbedWidget.ts [hunk 9]
```diff
@@ -14,22 +13,47 @@ const _log = debug("squatch-js:EMBEDwidget");
  *
  */
 export default class EmbedWidget extends Widget {
-  element: Element;
+  element: HTMLElement;
 
-  constructor(params: Params, selector = "#squatchembed") {
+  constructor(params: Params, container: string | HTMLElement | undefined) {
```
Comment (DerekSiemens) [inline src/widgets/EmbedWidget.ts:18]: Now that the behaviour exists for the embed widget, we should likely add it for the popup widget as well. If I am remembering right, it was left in the same state as this code was previously

File: package.json [hunk 10]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@saasquatch/squatch-js",
-  "version": "2.3.1",
+  "version": "2.3.2-29",
```
Comment (00salmon) [inline package.json:3]: 2.3.2 or 2.4.0?

Comment (DerekSiemens) [reply package.json:3]: 2.4.0 @00salmon

File: src/types.ts [hunk 11]
```diff
@@ -18,6 +19,8 @@ export interface WidgetConfig {
   user: User; // The user details
   widgetType?: WidgetType; // The content of the widget.
   engagementMedium?: EngagementMedium; // How to display the widget.
+  container?: HTMLElement; // Element to load the widget into.
+  trigger?: string; // Trigger element selector for opening the popup widget
```
Comment (00salmon) [inline src/types.ts:23]:
```suggestion
  trigger?: string; // Trigger element's selector for opening the popup widget
```

File: src/widgets/EmbedWidget.ts [hunk 12]
```diff
@@ -14,22 +13,48 @@ const _log = debug("squatch-js:EMBEDwidget");
  *
  */
 export default class EmbedWidget extends Widget {
-  element: Element;
+  element: HTMLElement;
 
-  constructor(params: Params, selector = "#squatchembed") {
+  constructor(params: Params, container?: string | HTMLElement) {
```
Comment (00salmon) [inline src/widgets/EmbedWidget.ts:18]:
```suggestion
  constructor(params: Params, container?: HTMLElement | string) {
```

File: src/widgets/EmbedWidget.ts [hunk 13]
```diff
@@ -14,22 +13,48 @@ const _log = debug("squatch-js:EMBEDwidget");
...
+      _log("loading widget with container", element);
+      // find element on page
+    } else {
+      element =
+        document.querySelector("#squatchembed") ||
+        document.querySelector(".squatchembed");
+      _log("loading widget with default selector", element);
+    }
 
     if (!element)
```
Comment (00salmon) [inline src/widgets/EmbedWidget.ts:39]: check for instance of HTMLElement

File: src/widgets/EmbedWidget.ts [hunk 14]
```diff
@@ -14,22 +13,48 @@ const _log = debug("squatch-js:EMBEDwidget");
...
+
+    // selector is a string
+    if (typeof container === "string") {
+      element = document.querySelector(container);
+      _log("loading widget with selector", element);
+      // selector is an HTML element
+    } else if (container) {
+      element = container as HTMLElement;
+      _log("loading widget with container", element);
+      // find element on page
```
Comment (00salmon) [inline src/widgets/EmbedWidget.ts:31]: check instance of, add warning if garbage is passed as container

**Review by loganvolkers — DISMISSED** [review]
Looks good to me

File: package.json [hunk 15]
```diff
@@ -44,7 +44,7 @@
     "static": "serve dist",
     "parcel": "cd demo && parcel index.html",
     "parcel:build": "cd demo && parcel build index.html",
-    "parcel:deploy": "surge demo/dist -d squathjs-demo.surge.sh"
+    "parcel:deploy": "surge demo/dist -d squathjs-demo-qa.surge.sh"
```
Comment (DerekSiemens) [inline package.json:47]:
```suggestion
    "parcel:deploy": "surge demo/dist -d squathjs-demo.surge.sh"
```

**Review by DerekSiemens — APPROVED** [review]

**Pull request merged**